### PR TITLE
feat: test-suite

### PR DESCRIPTION
**Pull request by sichanyoo**
## Issue \#

https://github.com/awslabs/aws-sdk-swift/issues/1148

## Description of changes

- Add mock implementations to use for unit tests
- Add unit tests for auth scheme middleware & signing middleware (the two new middlewares added as part of SRA I&A)
- Add CloudFront KeyValueStore as another service that uses rules-based auth scheme resolver customization
- Fix setting `SelectedAuthScheme` value in the `context : HttpContext` from building brand new one, to modifying original one via reference. Addresses an error where streaming flow does not have access to selected auth scheme because streaming signing flow is not part of the middleware stack & uses the original context.

## Scope

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Tests/ClientRuntimeTests/ClientRuntimeTests/NetworkingTests/Http/MiddlewareTests/AuthSchemeMiddlewareTests.swift [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    // Test exception cases
+    func testNoAuthSchemeResolverConfigured() async throws {
+        contextBuilder.attributes.remove(key: AttributeKeys.authSchemeResolver)
+        contextBuilder.withOperation(value: "fillerOp")
+        do {
+            try await AssertSelectedAuthSchemeMatches(builtContext: contextBuilder.build(), expectedAuthScheme: "")
+        } catch ClientError.authError(let message) {
+            XCTAssertEqual(message, "No auth scheme resolver has been configured on the service.")
+        } catch {
+            XCTFail("Expected exception was not thrown.")
```
Comment (jbelkins) [inline Tests/ClientRuntimeTests/ClientRuntimeTests/NetworkingTests/Http/MiddlewareTests/AuthSchemeMiddlewareTests.swift:34]: Maybe log the actual error / message here to let the developer know what was actually thrown.

Comment (jbelkins) [reply Tests/ClientRuntimeTests/ClientRuntimeTests/NetworkingTests/Http/MiddlewareTests/AuthSchemeMiddlewareTests.swift:34]: (Same in other tests where this pattern is used)

File: Tests/ClientRuntimeTests/ClientRuntimeTests/NetworkingTests/Http/MiddlewareTests/AuthSchemeMiddlewareTests.swift [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+    func testNoIdentityResolverConfigured() async throws {
+        contextBuilder.attributes.remove(key: AttributeKeys.identityResolvers)
+        contextBuilder.withOperation(value: "fillerOp")
+        do {
+            try await AssertSelectedAuthSchemeMatches(builtContext: contextBuilder.build(), expectedAuthScheme: "")
+        } catch ClientError.authError(let message) {
+            XCTAssertEqual(message, "No identity resolver has been configured on the service.")
+        } catch let error {
+            print(error)
```
Comment (jbelkins) [inline Tests/ClientRuntimeTests/ClientRuntimeTests/NetworkingTests/Http/MiddlewareTests/AuthSchemeMiddlewareTests.swift:46]: remove the `print()`

File: Tests/ClientRuntimeTests/ClientRuntimeTests/NetworkingTests/Http/MiddlewareTests/AuthSchemeMiddlewareTests.swift [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    func testNoIdentityResolverConfigured() async throws {
+        contextBuilder.attributes.remove(key: AttributeKeys.identityResolvers)
+        contextBuilder.withOperation(value: "fillerOp")
+        do {
+            try await AssertSelectedAuthSchemeMatches(builtContext: contextBuilder.build(), expectedAuthScheme: "")
+        } catch ClientError.authError(let message) {
+            XCTAssertEqual(message, "No identity resolver has been configured on the service.")
+        } catch let error {
```
Comment (jbelkins) [inline Tests/ClientRuntimeTests/ClientRuntimeTests/NetworkingTests/Http/MiddlewareTests/AuthSchemeMiddlewareTests.swift:45]: Just `} catch {` works.  Error is assigned to `error` by default.

**Review by jbelkins — CHANGES_REQUESTED** [review]
Just little stuff

File: Sources/ClientRuntime/Networking/Http/Middlewares/AuthSchemeMiddleware.swift [hunk 4]
```diff
@@ -100,8 +100,7 @@ public struct AuthSchemeMiddleware<OperationStackOutput: HttpResponseBinding,
         }
 
         // Set the selected auth scheme in context for subsequent middleware access, then pass to next middleware in chain
-        return try await next.handle(
-            context: context.toBuilder().withSelectedAuthScheme(value: selectedAuthScheme).build(), input: input
-        )
+        context.attributes.set(key: AttributeKeys.selectedAuthScheme, value: selectedAuthScheme)
+        return try await next.handle(context: context, input: input)
```
[the remark this answers is not in the document]
Comment (sichanyoo) [reply Sources/ClientRuntime/Networking/Http/Middlewares/AuthSchemeMiddleware.swift:104]: Code before change was also written by me, as this middleware is new in this project, lol. And yeah, that was the reason for dividing it into 2 lines haha

File: Sources/ClientRuntime/Networking/Http/Middlewares/ContentMD5Middleware.swift [hunk 5]
```diff
@@ -20,13 +20,11 @@ public struct ContentMD5Middleware<OperationStackOutput>: Middleware {
 
         switch input.body {
         case .data(let data):
-            guard
-                let data = data,
-                let bodyString = String(data: data, encoding: .utf8)
-            else {
+            guard let data = data else {
```
[the remark this answers is not in the document]
Comment (sichanyoo) [reply Sources/ClientRuntime/Networking/Http/Middlewares/ContentMD5Middleware.swift]: Changed as suggested.

File: Tests/ClientRuntimeTests/ClientRuntimeTests/NetworkingTests/Http/MiddlewareTests/AuthSchemeMiddlewareTests.swift (hunk 1 above)
Comment (sichanyoo) [reply Tests/ClientRuntimeTests/ClientRuntimeTests/NetworkingTests/Http/MiddlewareTests/AuthSchemeMiddlewareTests.swift:34]: Changed as suggested for several tests that this suggestion applies to.

File: Tests/ClientRuntimeTests/ClientRuntimeTests/NetworkingTests/Http/MiddlewareTests/AuthSchemeMiddlewareTests.swift (hunk 3 above)
Comment (sichanyoo) [reply Tests/ClientRuntimeTests/ClientRuntimeTests/NetworkingTests/Http/MiddlewareTests/AuthSchemeMiddlewareTests.swift:45]: Changed as suggested.

**Review by jbelkins — APPROVED** [review]

**Review by dayaffe — APPROVED** [review]

**Pull request merged**